### ISSUE TITLE
Retrieve code `length`

### ISSUE THREAD

**Issue by han0110** [appliedzkp/zkevm-specs#102]
Currently we can use lookup with input `(hash, index, byte, is_code)` to `bytecode_table` to retrieve the `byte` at specific `index`. However, we haven't taken the `length` of code into too much consideration, which is required for several places:

1. `CODESIZE`, `EXTCODESIZE`
2. `PUHS*`, `CODECOPY`, `EXTCODECOPY` right padding

    When copying out of code bound, implicit `STOP`s will be copied. It could be seen as right padding of the source. <br>See _yellowpaper[^yellowpaper] PUSH* and *CODECOPY_.

[^yellowpaper]: https://ethereum.github.io/yellowpaper/paper.pdf

3. Execution implicit `STOP`

    When execution has program counter out of code bound, an implicit `STOP` will be returned. See _yellowpaper (148)_.

Since the `length` already exists in Bytecode circuit, we can just also populate it into `bytecode_table` to be a 5-columns table <code>(hash, <b>length</b>, index, byte, is_code)</code>.

Then there is a caveat: If the code is empty, it's not allowed to be in current Bytecode circuit. So when we want to retrieve code `length` by `bytecode_table` lookup, we need to first check if `hash != EMPTY_HASH`.

Any feedback is appreciated.

**Comment by ed255** [comment]
I think the right padding behavior also happens when accessing call data with `CALLDATALOAD` and `CALLDATACOPY`
- CALLDATALOAD: https://github.com/ethereum/go-ethereum/blob/015fde9a2c/core/vm/instructions.go#L286
- CALLDATACOPY: https://github.com/ethereum/go-ethereum/blob/015fde9a2c/core/vm/instructions.go#L312
- getData definition: https://github.com/ethereum/go-ethereum/blob/015fde9a2c/core/vm/common.go#L54

In that case we do have `CallContext` -> `CallDataLength` in rw_table, but what's the strategy going to be?  This is how I imagined it: 2 lookups: 1st for length, and 2nd for value.  If query_length > lookup_length, then the 2nd value "output" will be shortcut with 0, and "ignore" the second lookup (maybe by looking up a known constant row).

So what about doing the same approach with bytecode by adding a tag?
- (hash, TagLength, length, 0, 0)
- (hash, TagByte, index, byte)
And then we always insert a row with (EMPTY_HASH, TagLength, 0, 0, 0)

Does this make sense?

**Comment by han0110** [comment]
> This is how I imagined it: 2 lookups: 1st for length, and 2nd for value. If query_length > lookup_length, then the 2nd value "output" will be shortcut with 0, and "ignore" the second lookup (maybe by looking up a known constant row).

Yes. More precisely, for sources:

1. Bytecode in `bytecode_table`
2. Memory of caller/callee in `rw_table`
3. Calldata in `tx_table`

the strategy is similar, we do:

1. Lookup `length`, and get selectors `q_lookups` by `q_lookups = [i < t for i in range(min(offset + N, length))]`, where `N` is a fixed amount we decide to lookup for a step (or depends on opcode itself like `*` of `PUSH*`).
2. Enable lookups to by `q_lookups`.

For source 2. and 3., we already have tags to get their `length`. And this issue is trying to figure out a simplest way to add a way to get bytecode's `length`.

> So what about doing the same approach with bytecode by adding a tag?

Adding a tag makes sense to me! I got stuck on how to have the empty code also in such table, but just like you said, we can design the circuit to always insert row `(EMPTY_HASH, BytecodeTableTag::Length, 0, 0, 0)` to make it work.

Naively the constraint might look like:

```python
# Variable `curr` and `next` are rows with `(hash, tag, val1, val2, val3)`
# When tag is Length, val1 == 0, val2 == 0, val3 == length.
# When tag is Byte, val1 == index, val2 == is_code, val3 == byte.

if curr.tag is Length:
    if curr.length == 0:
        assert next.tag == Length
    else:
        assert next.tag == Byte
        assert next._length == curr.length
else:
    ... # current constraints
```

**Comment by DreamWuGit** [comment]
@roynalnaruto starts to do `CODECOPY ` circuit implementation work, can have a look !

**Comment by roynalnaruto** [comment]
@han0110 I've added the above proposed additions with further constraints ([Refer this](https://github.com/appliedzkp/zkevm-specs/pull/148/commits/22955ef35a761a658cc160996375ec255c063cfe))

**Comment by han0110** [comment]
Closed by #148
